### DISCUSSION TITLE
Collecting needs for a revamped documentation

### DISCUSSION THREAD

**Discussion by emmanuelbernard** [langchain4j/langchain4j#2255] [Ideas]
Hello everyone 👋,

tl;dr; we need your help to understand how to best improve langchain4j documentation, read on 

## Context

I have been discussing with a few users and some contributors and I hear at times that the langchain4j documentation can be better. This discussion is to help start that work. And the first step is to understand what people find missing or lacking in the documentation today.
From there I imagine we will find tactical work to improve specific areas and strategic work where we will want to overall the structure, bring new guides etc.

I've tried to structure a bit the feedback but any feedback is useful. My apologies if some questions overlap a bit too much.
The most important aspect is I want to help the community to understand where to bring the documentation from where it is today to a more useful tool for beginners and experts alike.

With that in mind, I would love for you to spend 15 to 20 mins answering the questions below.

Many many thanks ! 🙏

## Ask 1 - Tactical improvements

When you found that Langchain4J doc was lacking,
What were you searching / looking for? 
Which area is hard for your to get data from?
What topic did you struggle with or still are struggling with?

## Ask 2 - General feedback

Where do you go to find documentation on langchain4j?

What part of the documentation do you like want want more of?

What part of the documentation is good as is?

What part of the documentation you struggle with and would like to see effort on?

What type of questions do you have when you hit the langchain4j doc?

## Ask 3 - Structure

Today the documentation is structured as such
- intro
- getting started
- tutorials
- integration
- useful materials
- examples (as a git repo)

What was your impression navigating the information?

Any additional section you would add or reorganise?

## Ask 4 - Open feedback

Anything you want to say?
And if you want to talk to a human, ping @emmanuelbernard and we can organise some chat, I would love to exchange with you.

**Comment by emmanuelbernard** [comment]
Note to me, there is a https://chat.langchain4j.dev/ that I don't see highlighted much

**Comment by michaelisvy** [comment]
hello, 
just to give some mixed (and a bit random) feedback. Hope that can be useful:

* When I come from Google and type "LangChain4J documentation", there is a landing page and I need to click on "Introduction" to access the doc. It's a bit misleading, because I might not know what is "Introduction", also I might be looking at advanced features. I would consider removing the landing page, so people go directly to this page: https://docs.langchain4j.dev/intro
* In the top-level chapters, I only have "Introduction", "Getting Started", "Tutorials", "Integrations".  It seems that all the interesting stuff is below the "Integrations" section. I would consider adding top level sections for most wanted items such as prompting, vector databases, etc.
* In many of the pages ([such as this one](https://docs.langchain4j.dev/integrations/language-models/anthropic)), code samples are incomplete, so I can't copy/paste directly from there. I need to browse the examples (and leave the documentation) in order to really picture how it works. It would be great to have everything in a single page.

Hope that helps :)

**Comment by emmanuelbernard** [comment]
One comment from Dan Dobrin in Discourse.


While the diataxis model mentioned by @glaforge is interesting, I'd like to add a different dimension here.
Documentation needs to expose a reader with information similar to Lego blocks: you build from the ground up and a reader can follow it step-by-step at the beginning, while, once gaining experience and building real apps, they come back and look up in the Lego stack for the info.

If I may say, in a very simple list, not a ToC:
What is Langchain4J
AI concepts
Where to start
Framework Architecture
Capabilities: mapping AI concepts to integrations (supported by the respective modules)
AI models (by type)
Cross-cutting concerns (guardrails, security)
Configuration (and auto-config)
Testing
Tutorials (separate tutorials from patterns)
Troubleshooting
Maintenance - upgrade notes, etc

For each one, start with diagrams, as images speak a thousand words. 
Say this is how guardrails work and here's the diagram to start, or here is adiagram with the lower level components and here is one for AI Services.

**Comment by emmanuelbernard** [comment]
@glaforge did mention the [Diataxis](https://diataxis.fr) framework to structure the doc. I was skeptical first but I like a few things in it. It does not have to be the alpha and the omega of your doc structure but is heavily helps to focus a given doc and not address concerns that should belong to another doc. So I'm thinking using it (at least partly) makes sense.

I wrote about my first impressions there in more details https://emmanuelbernard.com/blog/2024/12/19/diataxis/

**Comment by emmanuelbernard** [comment]
My next steps for info:
1. gather a bit more feedback if I can collect it
2. read the existing doc
3. start to come up with a structure proposal

**Comment by jdubois** [comment]
On my side, it's mostly having better description of how everything works. If I take specifically Azure OpenAI or GitHub Models, it's mostly me who needs time to write down everything cleanly.
Then, as long as the search engine works (which is the case) and that we have everything in markdown (so that my AI can understand it and help me afterwards), that's good enough.
Concerning AI, there are 2 points:
- LLMs will be trained on it at some point. For example, with JHipster, I can now ask "code me a REST Controller in the JHipster style". 
- Then, we can reference the markdown file in our IDE, so the AI assistant has access to the latest documentation and can help code with the latest version

I'm expecting that most of our users will use this documentation through their AI coding assistant, and that's why it's important to have good code snippets and clear explanations, so it pops up correctly in their IDE.

**Comment by glaforge** [comment]
Maybe the documentation should be available as [LLMs.txt](https://llmstxt.org/) (ie. this markdown single file format that contains the whole content of a site or documentation). We could just point the assistant to it.
